contracts/Comptroller.sol

```diff
@@ -1192,44 +1120,59 @@ contract Comptroller is ComptrollerV4Storage, ComptrollerInterface, ComptrollerE
+     */
+    function updateCompMarketIndex(address cToken, CompMarketState storage marketState, CompMarketState storage vestingState, bool isSupply, Exp memory marketBorrowIndex) internal {
+        uint compSpeed = compSpeeds[cToken];
         uint blockNumber = getBlockNumber();
-        uint deltaBlocks = sub_(blockNumber, uint(borrowState.block));
-        if (deltaBlocks > 0 && borrowSpeed > 0) {
-            uint borrowAmount = div_(CToken(cToken).totalBorrows(), marketBorrowIndex);
-            uint compAccrued = mul_(deltaBlocks, borrowSpeed);
-            Double memory ratio = borrowAmount > 0 ? fraction(compAccrued, borrowAmount) : Double({mantissa: 0});
-            Double memory index = add_(Double({mantissa: borrowState.index}), ratio);
-            compBorrowState[cToken] = CompMarketState({
-                index: safe224(index.mantissa, "new index exceeds 224 bits"),
-                block: safe32(blockNumber, "block number exceeds 32 bits")
-            });
+        uint deltaBlocks = sub_(blockNumber, uint(marketState.block));
+        if (deltaBlocks > 0 && compSpeed > 0) {
+            uint marketSize;
+            if (isSupply) {
+                marketSize = CToken(cToken).totalSupply();
+            } else {
+                marketSize = div_(CToken(cToken).totalBorrows(), marketBorrowIndex);
+            }
+            uint compAccrued;
+            Double memory ratio;
+            Double memory index;
+
+            if (lastVestingBlock > vestingState.block) {
+                uint deltaVestingBlocks = sub_(lastVestingBlock, uint(marketState.block));
+                compAccrued = mul_(deltaVestingBlocks, compSpeed);
+                ratio = marketSize > 0 ? fraction(compAccrued, marketSize) : Double({mantissa: 0});
+                // important reference to marketState index below as vesting index is not kept "up to date"
+                index = add_(Double({mantissa: marketState.index}), ratio);
+                vestingState.index = safe224(index.mantissa, "new index exceeds 224 bits");
+                vestingState.block = safe32(lastVestingBlock, "block number exceeds 32 bits");
+            }
+
+            compAccrued = mul_(deltaBlocks, compSpeed);
+            ratio = marketSize > 0 ? fraction(compAccrued, marketSize) : Double({mantissa: 0});
+            index = add_(Double({mantissa: marketState.index}), ratio);
+            marketState.index = safe224(index.mantissa, "new index exceeds 224 bits");
+            marketState.block = safe32(blockNumber, "block number exceeds 32 bits");
         } else if (deltaBlocks > 0) {
-            borrowState.block = safe32(blockNumber, "block number exceeds 32 bits");
+            if (lastVestingBlock > vestingState.block) {
+                vestingState.block = safe32(blockNumber, "block number exceeds 32 bits");
+            }
+            marketState.block = safe32(blockNumber, "block number exceeds 32 bits");
```

Comment:
Probably should check these are using the gas/ops (SSTORE, SLOAD) expected - we've found solc tends to do these types of assignments inefficiently